ADApp/ntndArrayConverterSrc/ntndArrayConverter_pvxs.cpp

```diff
@@ -0,0 +1,447 @@
+        case NDDataType_t::NDInt32:     {fromValue<int32_t>(src); break;};
+        case NDDataType_t::NDUInt32:    {fromValue<uint32_t>(src); break;};
+        case NDDataType_t::NDInt64:     {fromValue<int64_t>(src); break;};
+        case NDDataType_t::NDUInt64:    {fromValue<uint64_t>(src); break;};
+        case NDDataType_t::NDFloat32:   {fromValue<float_t>(src); break;};
+        case NDDataType_t::NDFloat64:   {fromValue<double_t>(src); break;};
+        default: {
+            throw std::runtime_error("invalid value data type");
+            break;
+        }
+    }
+}
+
+void NTNDArrayConverter::fromDimensions (NDArray *src) {
+    pvxs::shared_array<pvxs::Value> dims;
+    dims.resize(src->ndims);
+
+    for (int i = 0; i < src->ndims; i++) {
+        dims[i] = m_value["dimension"].allocMember();
+        dims[i].update("size", src->dims[i].size);
+        dims[i].update("offset", src->dims[i].offset);
+        dims[i].update("fullSize", src->dims[i].size);
+        dims[i].update("binning", src->dims[i].binning);
+        dims[i].update("reverse", src->dims[i].reverse);
+    }
+    m_value["dimension"] = dims.freeze();
+}
+
+void NTNDArrayConverter::fromDataTimeStamp (NDArray *src) {
+    double seconds = floor(src->timeStamp);
+    double nanoseconds = (src->timeStamp - seconds)*1e9;
+    // pvAccess uses Posix time, NDArray uses EPICS time, need to convert
+    seconds += POSIX_TIME_AT_EPICS_EPOCH;
+    m_value["dataTimeStamp.secondsPastEpoch"] = seconds;
+    m_value["dataTimeStamp.nanoseconds"] = nanoseconds;
+}
+
+void NTNDArrayConverter::fromTimeStamp (NDArray *src) {
+    // pvAccess uses Posix time, NDArray uses EPICS time, need to convert
+    m_value["timeStamp.secondsPastEpoch"] = src->epicsTS.secPastEpoch + POSIX_TIME_AT_EPICS_EPOCH;
+    m_value["timeStamp.nanoseconds"] = src->epicsTS.nsec;
+}
+
+template <typename valueType>
+void NTNDArrayConverter::fromAttribute (pvxs::Value dest_value, NDAttribute *src)
+{
+    valueType value;
+    src->getValue(src->getDataType(), (void*)&value);
+    dest_value["value"] = value;
+}
+
+void NTNDArrayConverter::fromStringAttribute (pvxs::Value dest_value, NDAttribute *src)
+{
+    const char *value;
+    src->getValue(src->getDataType(), (void*)&value);
+    dest_value["value"] = std::string(value);
+}
+
+void NTNDArrayConverter::fromAttributes (NDArray *src)
+{
+    NDAttributeList *srcList = src->pAttributeList;
+    NDAttribute *attr = NULL;
+    size_t i = 0;
+    pvxs::shared_array<pvxs::Value> attrs;
+    attrs.resize(src->pAttributeList->count());
+    while((attr = srcList->next(attr)))
+    {
+        NDAttrSource_t sourceType;
+        attr->getSourceInfo(&sourceType);
+        attrs[i] = m_value["attribute"].allocMember();
+        attrs[i].update("name", attr->getName());
+        attrs[i].update("descriptor", attr->getDescription());
+        attrs[i].update("source", attr->getSource());
+        attrs[i].update("sourceType", sourceType);
```

Comment:
fyi. `Value::update()` is define so that it can be chained.

```c++
attrs[i] = m_value["attribute"].allocMember()
.update("name", attr->getName())
.update("descriptor", attr->getDescription())
.update("source", attr->getSource())
.update("sourceType", sourceType);
```